infra/testing/sw-env.js

```diff
@@ -11,21 +11,158 @@
+// https://fetch.spec.whatwg.org/#request-class
+class Request {
+  constructor(url, options = {}) {
+    if (url instanceof Request) {
+      options = url;
+      url = options.url;
+    }
+
+    if (!url) {
+      throw new TypeError(`Invalid url: ${url}`);
+    }
+
+    this.url = url;
+    this.method = options.method || 'GET';
+    this.mode = options.mode || 'same-origin'; // FF defaults to cors
+    this.headers = new Headers(options.headers);
+
+    this.bodyUsed = !!options.body;
+    if (this.bodyUsed) {
+      this.body = options.body;
+    }
+  }
+
+  clone() {
+    return new Request(this.url, this);
+  }
+
+  text() {
+    return Promise.resolve(`${this.body || ''}`);
+  }
+}
+global.Request = Request;
+
+
+// Event
+// https://dom.spec.whatwg.org/#event
+class Event {
+  constructor(type, eventInitDict = {}) {
+    this.type = type;
+
+    this.bubbles = eventInitDict.bubbles || false;
+    this.cancelable = eventInitDict.cancelable || false;
+    this.composed = eventInitDict.composed || false;
+  }
+}
+global.Event = Event;
+
+// ExtendibleEvent
+// https://www.w3.org/TR/service-workers-1/#extendable-event
+class ExtendableEvent extends Event {
+  waitUntil(/* promise */) {
+    // TODO(philipwalton): keep track of the promises added to each event
+    // and expose them in some way so tests can assert logic doesn't run
+    // after all ExtendableEvents have settled.
+  }
+}
+global.ExtendableEvent = ExtendableEvent;
+
+// SyncEvent
+// https://wicg.github.io/BackgroundSync/spec/#sync-event
+class SyncEvent extends ExtendableEvent {
+  constructor(type, init = {}) {
+    super(type, init);
+
+    if (!init.tag) {
+      throw new TypeError(
+          `Failed to construct 'SyncEvent': required member tag is undefined.`);
+    }
+
+    this.tag = init.tag;
+    this.lastChance = init.lastChance || false;
+  }
+}
+global.SyncEvent = SyncEvent;
+
+// Fix incorrect addEventListener stub
+let listenerMap = new Map();
+
+global.addEventListener = (type, listener /* TODO: support `opts` */) => {
+  if (listenerMap.has(type)) {
+    listenerMap.get(type).add(listener);
+  } else {
+    listenerMap.set(type, new Set([listener]));
+  }
+};
+
+// dispatchEvent
+global.dispatchEvent = (event) => {
+  for (const listener of listenerMap.get(event.type)) {
+    listener(event);
+  }
+};
+
+global.__removeAllEventListeners = () => {
+  listenerMap = new Map();
+};
+
+// SyncManager
+// https://wicg.github.io/BackgroundSync/spec/#sync-manager-interface
+class SyncManager {
+  constructor() {
+    this._tagList = new Set();
+  }
+  async register(tagName) {
+    this._tagList.add(tagName);
+    return Promise.resolve();
+  }
+  async getTags() {
+    return Promise.resolve([...this._tagList]);
+  }
+}
+global.registration.sync = new SyncManager();
+
 
 // This is needed to ensure new URL('/', location), works.
-swEnv.location = 'https://example.com';
+global.location = 'https://example.com';
 
+// Needed to look like a real SW env
```

Comment:
Just an FYI - when you rebase, these next few lines will be removed (due to fix in service-worker-mock.

Reply:
Cool, rebased.